managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java

```diff
@@ -2015,6 +2020,59 @@ public void operationFailed(MetaStoreException e) {
                 });
     }
 
+    public void persistPendingAckPositionMetaInfo(PositionInfo.Builder pendingCumulativeAckMessagePositionBuilder,
+                                                  List<PendingAckMessageEntry> pendingAckMessageEntryBuilderList,
+                                                  String subName, MetaStoreCallback<Void> callback) {
+        if (isClosed()) {
+            callback.operationFailed(new MetaStoreException(
+                    new ManagedLedgerException.CursorAlreadyClosedException(name + " cursor already closed")));
+            return;
+        }
+
+        SubscriptionPendingAckMessages.Builder subscriptionPendingAckMessageBuilder = SubscriptionPendingAckMessages.newBuilder()
+                    .setPendingCumulativeAckMessagePosition(pendingCumulativeAckMessagePositionBuilder)
+                    .addAllPendingAckMessages(pendingAckMessageEntryBuilderList);
+
+        log.debug("[{}] [{}] [{}] Persisting subscription's pending ack messages to meta-data store",
+                ledger.getName(), name, subName);
+        ledger.getStore().asyncUpdateSubscriptionPendingAckMessages(ledger.getName(), name, subName,
+                pendingAckMessagesStoreStat, subscriptionPendingAckMessageBuilder.build(),
+                new MetaStoreCallback<Void>() {
+                    @Override
+                    public void operationComplete(Void result, Stat stat) {
+                        pendingAckMessagesStoreStat = stat;
+                        callback.operationComplete(result, stat);
+                    }
+
+                    @Override
+                    public void operationFailed(MetaStoreException e) {
+                        callback.operationFailed(e);
+                    }
+                });
+    }
+
+    public void getPendingAckPositionMetaInfo(String subName,
+                                              MetaStoreCallback<SubscriptionPendingAckMessages> callback) {
+        if (isClosed()) {
+            callback.operationFailed(new MetaStoreException(
+                    new ManagedLedgerException.CursorAlreadyClosedException(name + " cursor already closed")));
+            return;
+        }
+
+        ledger.getStore().asyncGetSubscriptionPendingAckMessages(ledger.getName(), name, subName,
```

Comment:
interesting... why do you need to create a new callback? Can't you just use the passed-in callback?

```
ledger.getStore().asyncGetSubscriptionPendingAckMessages(ledger.getName(), name, subName, callback);
```

Reply:
updated.